test/e2e/pkg/grammar/checker_test.go

```diff
@@ -17,88 +17,159 @@ var (
+	{[]*abci.Request{prepareProposal, processProposal, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{prepareProposal, processProposal, gotVote, extendVote, gotVote, finalizeBlock, commit}, true},
+	// proposer = *gotVote [prepare-proposal [process-proposal]] [extend]
+	{[]*abci.Request{gotVote, prepareProposal, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, prepareProposal, gotVote, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, prepareProposal, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, prepareProposal, gotVote, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, prepareProposal, processProposal, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, prepareProposal, processProposal, gotVote, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, prepareProposal, processProposal, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, prepareProposal, processProposal, gotVote, extendVote, gotVote, finalizeBlock, commit}, true},
+
+	// consensus-round = non-proposer
+	// non-proposer = *gotVote
+	// same as for proposer
+
+	// non-proposer = [process-proposal]
+	{[]*abci.Request{processProposal, finalizeBlock, commit}, true},
+	// non-proposer = [extend]
+	// same as for proposer
+
+	// non-proposer = *gotVote [process-proposal]
+	{[]*abci.Request{gotVote, processProposal, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, processProposal, finalizeBlock, commit}, true},
+	// non-proposer = *gotVote [extend]
+	// same as just [extend]
+
+	// non-proposer = [process-proposal] [extend]
+	{[]*abci.Request{processProposal, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{processProposal, gotVote, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{processProposal, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{processProposal, gotVote, extendVote, gotVote, finalizeBlock, commit}, true},
+
+	// non-proposer = *gotVote [prepare-proposal [process-proposal]] [extend]
+	{[]*abci.Request{gotVote, processProposal, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, processProposal, gotVote, extendVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, processProposal, extendVote, gotVote, finalizeBlock, commit}, true},
+	{[]*abci.Request{gotVote, gotVote, processProposal, gotVote, extendVote, gotVote, finalizeBlock, commit}, true},
+
+	{[]*abci.Request{prepareProposal, processProposal, processProposal, prepareProposal, processProposal, processProposal, processProposal, finalizeBlock, commit}, true},
+}
 
-func TestVerify(t *testing.T) {
-	tests := []struct {
-		name         string
-		abciCalls    []*abci.Request
-		isCleanStart bool
-		result       bool
-	}{
+func TestVerifyCleanStart(t *testing.T) {
```

Comment:
Why do we need to split it back into "clean start" and "recovery" ?

Reply:
Because a valid "clean start" execution is not necessarily a valid "recovery" execution and vice versa.

Reply:
I see. I trust your judgement here.
My question was motivated because we used to have the distinction between clean start and recovery in an earlier version (we even had two grammars), and then we managed to unify it. Hence I was curious why we're splitting this function whereas before this PR we only had `TestVerify`.

Reply:
The question is good. The thing is that since we now have one grammar, the parser will return that the execution is valid if it represents a valid clean start or recovery execution. In addition, it will return a parse tree. Consequently, to check that an execution is of a valid type also, I pass to the Verify method the execution type and then check if the parse tree matches the execution type. But this check is done inside the verify method; the parser alone cannot do this check since, for the grammar and the parser, the execution is valid if it satisfies any of the types.